impl/src/main/java/org/jboss/weld/bean/builtin/DecoratedBeanMetadataBean.java

```diff
@@ -50,6 +56,32 @@ protected void checkInjectionPoint(InjectionPoint ip) {
         }
     }
 
+    @Override
+    protected Bean<?> newInstance(InjectionPoint ip, CreationalContext<Bean<?>> ctx) {
+        checkInjectionPoint(ip);
+
+        WeldCreationalContext<?> decoratorContext = getParentCreationalContext(ctx);
+        WeldCreationalContext<?> beanContext = getParentCreationalContext(decoratorContext);
+        // in case of multiple decorators, the parent context can be another decorator
+        // TODO this is hack; instead I need to find where we create the CC hierarchy and why it is like this
```

Comment:
This "hack" seems reasonable to me ;-)

Reply:
Agreed, after some more investigation and discussion, I think this first approach I tried is actually a correct solution.
I will remove the TODO and keep just a comment explaining the need for it.